### PR TITLE
Attempt by Patrik Bóna

### PR DESCRIPTION
**Pull request by mrhead**
I am afraid that I am not able to do more refactoring, so here is my pull request.

### CODE REVIEW

File: output.txt [hunk 1]
```diff
@@ -0,0 +1,41 @@
+984 MACBETH
```
Comment (twosevenzero) [inline output.txt]: Even after fixing the bug in my code, I am not getting a number this high for Macbeth. Do we know what the actual answer is?

File: spec/shakespeare_analyzer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require 'spec_helper'
+require 'shakespeare_analyzer'
+require 'stringio'
+
+describe ShakespeareAnalyzer do
+  describe '#initialize' do
+    it 'reads provided URI and stores its content to @file_content' do
+      stub_request(:get, 'http://www.example.com/test_file.txt').to_return(body: 'This is just a test file!')
```
Comment (twosevenzero) [inline spec/shakespeare_analyzer_spec.rb:8]: I need to learn what these stubs are. Do they interrupt flow and substitute content?

File: spec/shakespeare_analyzer_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      analyzer = ShakespeareAnalyzer.new('http://www.example.com/test_file.txt')
+      expect(analyzer.instance_variable_get(:@file_content)).to eq 'This is just a test file!'
```
Comment (twosevenzero) [inline spec/shakespeare_analyzer_spec.rb:10]: I really could have used this method when I was writing my tests. "instance_variable_get"
Ill have to remember that for next time.

File: output.txt (hunk 1 above)
Comment (mrhead) [inline output.txt]: Well I do not know what are the correct numbers ;).

I perform my tests on small test.xml file and for this file everything is calculated correctly. But I do not know if I do not miss some edge cases in XML.

File: spec/shakespeare_analyzer_spec.rb (hunk 2 above)
Comment (mrhead) [inline spec/shakespeare_analyzer_spec.rb:8]: Yeah, check this gem: https://github.com/bblimke/webmock

I've found it when I was looking for solution how to test http download. It is awesome.

File: output.txt (hunk 1 above)
Comment (mrhead) [inline output.txt]: Well I just found bug in my code, I need to fix it. Hopefully then we will have same numbers ;).

Comment (mrhead) [inline output.txt]: Bug fixed, now we have same numbers for Macbeth ;)

**Comment by JESii** [comment]
Hi, Patrik... Nice work! You might want to consider updating the output so that the names aren't upper-cased.

File: bin/shakespeare_analyzer.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
+
+require_relative '../lib/shakespeare_analyzer.rb'
+
+DEFAULT_XML_URL = 'http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml'
+
+analyzer = ShakespeareAnalyzer.new(DEFAULT_XML_URL)
```
Comment (andyw8) [inline bin/shakespeare_analyzer.rb:7]: I think it would be good to avoid passing a URL to the initializer, and just provide the data instead. That means you could supply text from a file, and it should also make testing easier.

File: lib/shakespeare_analyzer.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def run
+    print_speakers_sorted_by_line_count
+  end
+
+  private
+
+  def get_content_from_uri(uri)
+    open(uri, proxy: ENV['http_proxy']).read
```
Comment (andyw8) [inline lib/shakespeare_analyzer.rb:16]: This suggests the Shakespeare class is doing too much - it's parsing the text, but also having to deal with the complexities of a proxy server.

File: spec/shakespeare_analyzer_spec.rb (hunk 3 above)
Comment (andyw8) [inline spec/shakespeare_analyzer_spec.rb:10]: I would actually recommend against using `instance_variable_get` in a test, because it means you're inspecting the internal state of the object, rather than testing it's external behaviour.

File: spec/test_files/test.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0"?>
```
Comment (andyw8) [inline spec/test_files/test.xml:1]: I like this concept of having the XML samples in a separate file. I embedded them within my spec, but it makes it very cluttered. I guess there's a trade-off though, in that you need to look at two places to understand what the spec is doing.

**Comment by mrhead** [comment]
@andyw8 thanks for all your comments. They all seems valid to me and I will keep them in mind for my next projects. Thanks!

File: bin/shakespeare_analyzer.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env ruby
+
+require_relative '../lib/shakespeare_analyzer.rb'
+
+DEFAULT_XML_URL = 'http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml'
```
Comment (r00k) [inline bin/shakespeare_analyzer.rb:5]: You're one of the only folks to extract this into a constant. I like it. :)

File: lib/shakespeare_analyzer.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
+require 'open-uri'
+require_relative 'xml_parser'
+
+class ShakespeareAnalyzer
+  def initialize(uri)
+    @file_content = get_content_from_uri(uri)
+  end
+
+  def run
+    print_speakers_sorted_by_line_count
```
Comment (r00k) [inline lib/shakespeare_analyzer.rb:10]: Nice high level of abstraction.

File: lib/shakespeare_analyzer.rb (hunk 5 above)
Comment (r00k) [inline lib/shakespeare_analyzer.rb:16]: Agreed with @andyw8.

File: spec/shakespeare_analyzer_spec.rb [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    end
+  end
+
+  describe '#run' do
+    it 'prints list of speakers sorted by line count' do
+      test_file = File.dirname(__FILE__) + '/test_files/test.xml'
+      stub_request(:get, 'http://www.example.com/test.xml').to_return(body: File.read(test_file))
```
Comment (r00k) [inline spec/shakespeare_analyzer_spec.rb:17]: Notice how you have to stub out the http library when what you really want to do is check that the right parsing/counting is happening. This is a strong hint that your class has a few too many responsibilities.

File: lib/shakespeare_analyzer.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  private
+
+  def get_content_from_uri(uri)
+    open(uri, proxy: ENV['http_proxy']).read
+  end
+
+  def print_speakers_sorted_by_line_count
```
Comment (shreve) [inline lib/shakespeare_analyzer.rb:19]: Nice job keeping even your private methods expressive, readable, and short. This whole file is really easy to understand, but I particularly like the clarity of this method.

File: spec/shakespeare_analyzer_spec.rb (hunk 9 above)
Comment (charlieanna) [inline spec/shakespeare_analyzer_spec.rb:17]: I never thought of stubbing the internet here. Thanks. Will add it to my solution too.
